ui/src/app/base/constants.ts

```diff
@@ -1,3 +1,10 @@
+export const ACTION_STATUS = {
+  error: "error",
+  idle: "idle",
+  loading: "loading",
+  success: "success",
+} as const;
```

Comment:
There are some cases where using a mutually exclusive status might not reflect the state of the data in Redux. For example `discovery.list` does not get cached so when we first fetch the list of discoveries we'd end up with `fetch: "success"` then later on if we tried to fetch the list again it would go in to `fetch: "loading"` and if it failed we'd have `fetch: "error"` but really the data still exists in the store and we should display it, we just couldn't update it.

It's possible that if we consider the status just the status of current request and stop using it to decide if data should be displayed then in might be OK (e.g. at the moment we hide content and display a spinner when it's in loading state, but instead we'd hav to check to see if there was data in the store).

We probably need to look into that a bit more to see if it'll cause us problems.

Reply:
Thanks for pointing this out.

What I had in mind when I suggested an enum status was to model the state like a finite-state machine. What this means here - we might just be missing a status for the refetch error state? 

```tsx
export const ACTION_STATUS = {
  idle: "idle",
  loading: "loading",
  loadingError: "loadingError",
  refetchError: "refetchError",
  success: "success",
} as const;
```

Alternatively, we could add a common field for storing the number of retries that'd indicate if we're dealing with an initial load or refetch error. I'm a fan of how `xState` deals with this problem. Pasting an example, I hope it helps illustrate what I had in mind.

```tsx
import { createMachine, assign } from 'xstate';

interface Context {
  retries: number;
}

const fetchMachine = createMachine<Context>({
  id: 'fetch',
  initial: 'idle',
  context: {
    retries: 0
  },
  states: {
    idle: {
      on: {
        FETCH: 'loading'
      }
    },
    loading: {
      on: {
        RESOLVE: 'success',
        REJECT: 'failure'
      }
    },
    success: {
      type: 'final'
    },
    failure: {
      on: {
        RETRY: {
          target: 'loading',
          actions: assign({
            retries: (context, event) => context.retries + 1
          })
        }
      }
    }
  }
});
```
https://stately.ai/viz